### PR TITLE
Ltd 4585 add celery

### PR DESCRIPTION
**Pull request by Tllew**

### CODE REVIEW

File: Procfile [hunk 1]
```diff
@@ -1,3 +1,4 @@
 # See gunicorn.conf.py for more configuration.
 web: python manage.py migrate && gunicorn conf.wsgi:application
 worker: python manage.py process_tasks --log-std
+celeryworker: celery -A conf worker -l info
```
Comment (kevincarrogan) [inline Procfile:4]: Should we also have a celery beat worker here too?

File: docker-compose.yml [hunk 2]
```diff
@@ -30,12 +32,61 @@ services:
...
+  celery:
+    build: .
+    volumes:
+      - .:/app
+    env_file: .env
+    links:
+      - lite-hmrc-postgres
+      - redis
+    depends_on:
+      - lite-hmrc-intg
```
Comment (kevincarrogan) [inline docker-compose.yml:57]: It's interesting that we have the dependency this way around, I would imagine that we'd want to say that the main service depends on this being up.

File: docker-compose.yml [hunk 3]
```diff
@@ -30,12 +32,61 @@ services:
...
+    expose:
+      - 6379
+    ports:
+      - 6379:6379
+    networks:
+      - lite
 
+networks:
+  lite:
+    external: true
```
Comment (kevincarrogan) [inline docker-compose.yml:90]: I don't think we need to setup an external network here.

File: core/tests/test_celery_tasks.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+from django.conf import settings
+from django.utils import timezone
+from rest_framework.test import APITestCase
+
+from core import celery_tasks
+from mail.enums import ReplyStatusEnum
+from mail.models import Mail
+
+
+class FlagsUpdateTest(APITestCase):
```
Comment (kevincarrogan) [inline core/tests/test_celery_tasks.py:12]: I think this test needs to be renamed.

File: conf/settings.py [hunk 5]
```diff
@@ -255,3 +259,34 @@
 AZURE_AUTH_TENANT_ID = env.str("AZURE_AUTH_TENANT_ID")
 
 SEND_REJECTED_EMAIL = env.bool("SEND_REJECTED_EMAIL", default=True)
+
+DEFAULT_AUTO_FIELD = "django.db.models.AutoField"
+
+if VCAP_SERVICES:
+    if "aws-s3-bucket" not in VCAP_SERVICES:
+        raise Exception("S3 Bucket not bound to environment")
```
Comment (kevincarrogan) [inline conf/settings.py:267]: I think you might have copied and pasted too much from the `lite-api` as we don't need to do the bucket check for this service.

Comment (kevincarrogan) [reply conf/settings.py:267]: It's actually probably a good idea to push this onto a test environment as I _think_ you would have seen this break quite quickly once you deploy it.

File: Procfile (hunk 1 above)
Comment (currycoder) [reply Procfile:4]: I think that would be handy while we are here

File: core/tests/test_celery_tasks.py (hunk 4 above)
Comment (currycoder) [reply core/tests/test_celery_tasks.py:12]: 👀 https://github.com/uktrade/lite-api/pull/1050/files#diff-e2b82cb4633cc863bbb13ebd2337311d8818306814adf656cf6d8e3c157d6953R7

File: conf/settings.py (hunk 5 above)
Comment (Tllew) [reply conf/settings.py:267]: Ahh oops, I thought I took all the aws-s3 references out of settings.py

File: docker-compose.yml (hunk 3 above)
Comment (Tllew) [reply docker-compose.yml:90]: I think this was something I added in the desperate attempts to get rid of an ampq:// connection interrupted error I was having, but I needed to `cp local.env .env`

File: Procfile (hunk 1 above)
Comment (Tllew) [reply Procfile]: This has now been added

File: conf/settings.py [hunk 6]
```diff
@@ -255,3 +259,34 @@
...
+if REDIS_BASE_URL:
+    # Give celery tasks their own redis DB - future uses of redis should use a different DB
+    REDIS_CELERY_DB = env("REDIS_CELERY_DB", default=0)
+    is_redis_ssl = REDIS_BASE_URL.startswith("rediss://")
+    url_args = {"ssl_cert_reqs": "CERT_REQUIRED"} if is_redis_ssl else {}
+
+    CELERY_BROKER_URL = _build_redis_url(REDIS_BASE_URL, REDIS_CELERY_DB, **url_args)
+    CELERY_RESULT_BACKEND = CELERY_BROKER_URL
+
+CELERY_ALWAYS_EAGER = env.bool("CELERY_ALWAYS_EAGER", False)
```
[the remark this answers is not in the document]
Comment (Tllew) [reply conf/settings.py]: I've removed `CELERY_ALWAYS_EAGER` so we'll find out during testing

File: celerybeat-schedule
[the remark this answers is not in the document]
Comment (Tllew) [reply celerybeat-schedule:1]: Done

**Review by saruniitr — APPROVED** [review]
with the changes verified on an env, this looks good to me

**Review by currycoder — APPROVED** [review]

**Review by kevincarrogan — APPROVED** [review]

**Pull request merged**